substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrChunkWriter.java

```diff
@@ -197,47 +267,120 @@ public void patchFileHeader(SignedWord constantPoolPosition, SignedWord metadata
+        } else {
+            getFileSupport().writeByte(fd, COMPLETE);
+        }
+        getFileSupport().writeByte(fd, (byte) 0);
+        short flags = 0;
+        flags += compressedInts ? 1 : 0;
+        flags += isFinal ? 1 * 2 : 0;
+
+        getFileSupport().writeShort(fd, flags);
+
+        // need to move pointer back to correct position for next write
+        getFileSupport().seek(fd, currentPos);
+    }
+
+    private byte nextGeneration() {
+        if (generation == MAX_BYTE) {
+            // similar to Hotspot, restart counter if required.
+            generation = 1;
+            return MAX_BYTE;
+        }
+        return generation++;
+    }
+
+    private SignedWord writeThreadCheckpointEvent(JfrConstantPool threadRepo, boolean flush) {
+
+        SignedWord start = beginEvent();
+
+        if (lastCheckpointOffset < 0) {
+            lastCheckpointOffset = start.rawValue();
+        }
+        writeCompressedLong(CONSTANT_POOL_TYPE_ID);
+        writeCompressedLong(JfrTicks.elapsedTicks());
+        writeCompressedLong(0); // duration
+        writeCompressedLong(lastCheckpointOffset - start.rawValue()); // deltaToNext
+        writeCompressedLong(8); // Checkpoint type is "THREADS"
+
+        SignedWord poolCountPos = getFileSupport().position(fd);
+        getFileSupport().writeInt(fd, 0); // We'll patch this later.
+
+        int poolCount = threadRepo.write(this, flush);
+
+        SignedWord currentPos = getFileSupport().position(fd);
+        getFileSupport().seek(fd, poolCountPos);
+        getFileSupport().writeInt(fd, makePaddedInt(poolCount));
+        getFileSupport().seek(fd, currentPos);
+        endEvent(start);
+        lastCheckpointOffset = start.rawValue();
+
+        return start;
     }
 
-    private SignedWord writeCheckpointEvent(JfrConstantPool[] repositories) {
+    private SignedWord writeCheckpointEvent(JfrConstantPool[] repositories, boolean flush) {
+        assert lock.isHeldByCurrentThread();
         SignedWord start = beginEvent();
+
+        if (lastCheckpointOffset < 0) {
+            lastCheckpointOffset = start.rawValue();
+        }
         writeCompressedLong(CONSTANT_POOL_TYPE_ID);
         writeCompressedLong(JfrTicks.elapsedTicks());
         writeCompressedLong(0); // duration
-        writeCompressedLong(0); // deltaToNext
+        writeCompressedLong(lastCheckpointOffset - start.rawValue()); // deltaToNext
         writeBoolean(true); // flush
 
         SignedWord poolCountPos = getFileSupport().position(fd);
         getFileSupport().writeInt(fd, 0); // We'll patch this later.
         JfrConstantPool[] serializers = JfrSerializerSupport.get().getSerializers();
-        int poolCount = writeConstantPools(serializers) + writeConstantPools(repositories);
+
+        int poolCount = 0;
+        if (newChunk) {
+            poolCount = writeConstantPools(serializers, flush);
+        }
+        poolCount += writeConstantPools(repositories, flush);
```

Comment:
Doing them separately should be ok I think
- Only one thread can be writing the repositories at a time because of the chunk writer lock.
- If the flushing thread is currently writing out the repositories and races with other threads adding entries to the repositories, it should not be an issue as long as we stick to the write order (so one entry cannot reference another that doesn't yet exist).

In Hotspot, it seems as though the repositories are locked and serialized individually too. Do you think it would be better to acquire all the repo locks before serializing the constant pools? This might have less of impact once we are only flushing new data (instead of how it's currently done).

Reply:
If HotSpot locks and serializes the repositories individually as well, then your current approach should be fine.